firedrake/slate/preconditioners.py

```diff
@@ -292,8 +297,9 @@ def update(self, pc):
         """Update by assembling into the operator. No need to
         reconstruct symbolic objects.
         """
-        self._assemble_S()
-        self.S.force_evaluation()
+        if self.mat_type == "aij":
+            self._assemble_S()
+            self.S.force_evaluation()
```

Comment:
This is wrong.  You need to assemble in the matrix-free case too.  That way, the PETSc KSP notices if the operator changes.